price-feeder/oracle/provider/kraken.go

```diff
@@ -23,100 +26,385 @@ type (
-		client:  newDefaultHTTPClient(),
+// NewKrakenProvider returns a new Kraken provider with the WS connection and msg handler.
+func NewKrakenProvider(ctx context.Context, logger zerolog.Logger, pairs ...types.CurrencyPair) (*KrakenProvider, error) {
+	wsURL := url.URL{
+		Scheme: "wss",
+		Host:   krakenHost,
+	}
+
+	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to Kraken websocket: %w", err)
+	}
+
+	provider := &KrakenProvider{
+		wsURL:           wsURL,
+		wsClient:        wsConn,
+		logger:          logger.With().Str("module", "oracle").Logger(),
+		tickers:         map[string]TickerPrice{},
+		subscribedPairs: map[string]types.CurrencyPair{},
+	}
+
+	if err := provider.SubscribeTickers(pairs...); err != nil {
+		return nil, err
+	}
+
+	go provider.handleWebSocketMsgs(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the saved map.
+func (p *KrakenProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
+	tickerPrices := make(map[string]TickerPrice, len(pairs))
+
+	for _, cp := range pairs {
+		key := cp.String()
+		tickerPrice, ok := p.tickers[key]
+		if !ok {
+			return nil, fmt.Errorf("kraken provider failed to get ticker price for %s", key)
+		}
+		tickerPrices[key] = tickerPrice
+	}
+
+	return tickerPrices, nil
+}
+
+// SubscribeTickers subscribe to all currency pairs and
+// add the new ones into the provider subscribed pairs.
+func (p *KrakenProvider) SubscribeTickers(cps ...types.CurrencyPair) error {
+	pairs := make([]string, len(cps))
+
+	for i, cp := range cps {
+		pairs[i] = currencyPairToKrakenPair(cp)
+	}
+
+	if err := p.subscribePairs(pairs...); err != nil {
+		return err
 	}
+
+	p.setSubscribedPairs(cps...)
+	return nil
 }
 
-func NewKrakenProviderWithTimeout(timeout time.Duration) *KrakenProvider {
-	return &KrakenProvider{
-		baseURL: krakenBaseURL,
-		client:  newHTTPClientWithTimeout(timeout),
+// handleWebSocketMsgs receive all the messages from the provider
+// and controls to reconnect the web socket.
+func (p *KrakenProvider) handleWebSocketMsgs(ctx context.Context) {
+	reconnectTicker := time.NewTicker(defaultMaxConnectionTime)
+	defer reconnectTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewWSMessage):
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message
+				p.logger.Err(err).Msg("Kraken provider could not read message")
+				if err := p.ping(); err != nil {
+					p.logger.Err(err).Msg("failed to send ping")
+					p.keepReconnecting()
+				}
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz)
+
+		case <-reconnectTicker.C:
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("Kraken provider receved reconnecting")
+				p.keepReconnecting()
```

Comment:
Same comment as above :)

Reply:
what if it can't reconnect? like we discussed [here](https://github.com/umee-network/umee/pull/551#discussion_r808424921)

So, I tried to have a max of tries when `keepReconnecting` by [3 times](https://github.com/umee-network/umee/pull/580/commits/677b6e0832a535dfb7dee6605028421aa2ff1e94) 
What do you think?

Reply:
Ahh nevermind, I think this is OK. Because `keepReconnecting` will block until it exits, meaning we never loop over the main ticker here again until that exists. So I think we're fine 👍